build-logic/uber-plugins/src/main/kotlin/gradlebuild.gradle-platform.gradle.kts

```diff
@@ -0,0 +1,54 @@
+import org.gradle.api.artifacts.component.ProjectComponentIdentifier
+import org.gradle.api.attributes.Category
+import org.gradle.api.attributes.TestSuiteType
+import org.gradle.api.attributes.VerificationType
+import org.gradle.kotlin.dsl.*
+
+/*
+ * Copyright 2023 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/*
+ * Should be applied to projects that represent a Gradle platform.
+ * This plugin adds tasks that aggregate actions across all subprojects of the platform.
+ * This plugin should eventually be expanded to handle other platform-scope respoonsibilities,
+ * for example distribution/packaging.
+ */
+
+val subproject = configurations.dependencyScope("subproject")
+
+val resolvedSubprojects = configurations.resolvable("resolvedSubprojects") {
+    extendsFrom(subproject.get())
+    // We do not want to depend on the dependencies of our subprojects.
+    // The intention is to only run the tests of this platform only.
+    isTransitive = false
+}
+
+tasks.register("test") {
+    dependsOn(resolvedSubprojects.get().incoming.artifactView {
+        withVariantReselection()
+        componentFilter { it is ProjectComponentIdentifier }
+        attributes {
+            attribute(Category.CATEGORY_ATTRIBUTE, objects.named(Category::class.java, Category.VERIFICATION))
+            attribute(TestSuiteType.TEST_SUITE_TYPE_ATTRIBUTE, objects.named(TestSuiteType::class.java, "unit-test"))
+            attribute(VerificationType.VERIFICATION_TYPE_ATTRIBUTE, objects.named(VerificationType::class.java, VerificationType.TEST_RESULTS))
+        }
+    }.files)
+}
+
+// In order to support embeddedIntegTest, etc, we would need to use test suites to implement
```

Comment:
💭 Shouldn't we just make new lifecycle tasks that depend on the test/integ test tasks for now and add task dependencies directly?  What does wiring inputs properly buy us IRT the use case of running tests within a platform?

Reply:
I think that could be an option, but to me depending on tasks directly from another project feel like too much of "digging into internals" of the other project. 

I'm not completely against doing it, but I would prefer an option that uses dependency-management for cross-project dependencies

Reply:
I think I'd usually agree, except in this case we're also adding a contrived input dependency - we don't really care about or need to obtain the test results here, we just want to kick off the tasks.  This seems like a task dependency is actually the more proper solution for a lifecycle task like this.

Could the platform convention plugin "export" the test tasks from the subprojects via lazily adding then to a property on an extension?  

```
platforms {
	jvm {
		testTasks = ... // ListProvider of tasks
		integTestTasks = ...  // ListProvider of tasks
        ...
    }

    software {
		testTasks = ... // ListProvider of tasks
		integTestTasks = ...  // ListProvider of tasks
        ...
    }
}
```

Then we don't need to reach into the projects directly to add the dependencies, only to these extension properties.

Reply:
I agree we don't really necessarily want the test results from the other projects, just that we want the other projects tests to run. 

I'm not completely sure why I am not a fan of direct task dependencies between projects. This feels to me like it goes against the Isolated Projects philosophy, even if it would still technically be compatible. 

Perhaps this requires a higher level discussion on how we expect different projects to "talk" to each other